contributors/design-proposals/federated-deployment.md

```diff
@@ -0,0 +1,88 @@
+
+## Implementation
+
+The implementation of Federated Deployment Controller is largely based on the existing Federated ReplicaSet Controller. The following changes however will be added (on top of obvious changes like “s/replicaset/deployment/”):
+
++ For each of the pod templates, Federated Deployment will make sure that puppetized FederatedReplicaset with the name deployment_name + “-” + pod_template_hash is created, with the appropriate revision number.
+
++ Local deployment checks will occur according to annotation (if present) or alphabetically.
++ If Federated Deployment with `strategy == RollingUpdate` spots a Local Deployment with different pod template it updates its only if there is no other update going on. Updates are spotted by checking if:
+    + `deployment.status.updatedReplicas == deployment.status.replicas`
+    + `deployment.status.unavailableReplicas == 0`
+    + `deployment.status.observed_generation == deployment.metadata.generation`
+
+  The Local Deployment will get a chance to be updated at a later time. Every change in Local Deployments trigger a Federated Deployment reconciliation. So once one Local Deployment is done with the update and publishes its new state the reconciliation will be triggered and next Local Deployment update started.
++ Federated Deployments with `strategy == Recreate` update all clusters at once.
++ FederatedDeployments always update replica numbers (if needed), no matter whether some update is going on or not.
++ Rollback REST api is registered.
++ Federated Deployment Controller has support for deleting old Federated ReplicaSets (just like in regular deployments). Replicasets are deleted in a non-cascading way.
+
+Moreover a “puppetized” Federated ReplicaSet (PFRS) support will be added to Federated Replicaset Controller:
+
+ + PFRS are marked with an annotation. The presence of the annotation makes the FRS a puppet.
+ + PFRS are created by Federated Deployments and contain the proper pod spec.
+ + PFRS monitors all underlying clusters where the replicas could potentially be found and updates their statuses and spec. Replicas based on the underlying cluster content.
+ + PFRS never updates RS in underlying clusters.
+
+## Tests
+
+Apart from similar tests that should be done for FRS, FD tests need to check:
+ + If a failure in update stops the whoe update.
+ + If a new update is requested before the previous one was completed.
+ + If all of the rollout mechanics and kubectl command work
+ + If Federated ReplicaSets have the correct statuses and replica counts. 
+
```

Comment:
Also:
* Make sure deletion works as expected: FRS is not deleted unless FD is deleted and is deleted when FD is deleted.
* federated deployment controller overrites any direct updates to PFRS by users.